rand_distr/tests/cdf.rs

```diff
@@ -195,3 +555,434 @@ fn binomial() {
         });
     }
 }
+
+#[test]
+fn geometric() {
+    fn cdf(k: i64, p: f64) -> f64 {
+        if k < 0 {
+            0.0
+        } else {
+            1.0 - (1.0 - p).powi(1 + k as i32)
+        }
+    }
+
+    let parameters = [0.3, 0.5, 0.7, 0.0000001, 0.9999];
+
+    for (seed, p) in parameters.into_iter().enumerate() {
+        let dist = rand_distr::Geometric::new(p).unwrap();
+        test_discrete(seed as u64, dist, |k| cdf(k, p));
+    }
+}
+
+#[test]
+fn hypergeometric() {
+    fn cdf(x: i64, n: u64, k: u64, n_: u64) -> f64 {
+        let min = if n_ + k > n { n_ + k - n } else { 0 };
+        let max = k.min(n_);
+        if x < min as i64 {
+            return 0.0;
+        } else if x >= max as i64 {
+            return 1.0;
+        }
+
+        (min..x as u64 + 1).fold(0.0, |acc, k_| {
+            acc + (ln_binomial(k, k_) + ln_binomial(n - k, n_ - k_) - ln_binomial(n, n_)).exp()
+        })
+    }
+
+    let parameters = [
+        (15, 13, 10),
+        (25, 15, 5),
+        (60, 10, 7),
+        (70, 20, 50),
+        (100, 50, 10),
+        // (100, 50, 49), // Fail case
+    ];
+
+    for (seed, (n, k, n_)) in parameters.into_iter().enumerate() {
+        let dist = rand_distr::Hypergeometric::new(n, k, n_).unwrap();
+        test_discrete(seed as u64, dist, |x| cdf(x, n, k, n_));
+    }
+}
+
+#[test]
+fn poisson() {
+    use rand_distr::Poisson;
+    fn cdf(k: i64, lambda: f64) -> f64 {
+        if k < 0 || lambda <= 0.0 {
+            return 0.0;
+        }
+
+        1.0 - gamma_lr(k as f64 + 1.0, lambda)
+    }
+    let parameters = [
+        0.1, 1.0, 7.5,
+        // 1e9, passed case but too slow
+        // 1.844E+19,  // fail case
+    ];
+
+    for (seed, lambda) in parameters.into_iter().enumerate() {
+        let dist = Poisson::new(lambda).unwrap();
+        test_discrete::<f64, Poisson<f64>, _>(seed as u64, dist, |k| cdf(k, lambda));
+    }
+}
+
+fn ln_factorial(n: u64) -> f64 {
+    (n as f64 + 1.0).lgamma().0
+}
+
+fn ln_binomial(n: u64, k: u64) -> f64 {
+    ln_factorial(n) - ln_factorial(k) - ln_factorial(n - k)
+}
+
+/// https://docs.rs/statrs/latest/src/statrs/function/gamma.rs.html#260-347
+fn gamma_lr(a: f64, x: f64) -> f64 {
```

Comment:
statrs is MIT-licenced, which permits modification and redistribution subject to the copyright notice and the licence itself being redistributed. We do the latter, but should include [the copyright notice](https://github.com/statrs-dev/statrs/blob/master/LICENSE.md). I think the best way to do this would be to push this method into a new file only for content copied from statrs. Alternatively, just copy the copyright line here (making clear it applies only to this function).

Secondly, the source of this method has some error checking. We don't need to return an error type, but should still `assert` those cases don't happen.